# Pandas

## Some Installation Issues

### Choose Kernel
![alt text](images/choose_kernel.jpg)

### SQLAlchemy ModuleNotFoundError: No module named 'psycopg2'
![alt text](images/pscylopg2_modulenotfound_error.png)

#### Understanding the Error:
The error ModuleNotFoundError: No module named `psycopg2` indicates that the `psycopg2` package, which is a PostgreSQL adapter for Python, is not installed in your Python environment.

#### Why is it required to have the `psycopg2` adapter installed for SQLAlchemy to work?
The PostgreSQL adapter `psycopg2` is needed because it provides a way for Python applications to interact with PostgreSQL databases. Here are the main reasons why `psycopg2` is necessary:
1. **Database Connectivity**
`psycopg2` is a PostgreSQL adapter, which means it acts as a bridge between Python and a PostgreSQL database. It allows Python applications to connect to a PostgreSQL database, execute SQL commands, and fetch results.

2. **Executing SQL Commands**
With `psycopg2`, you can run SQL commands like `SELECT`, `INSERT`, `UPDATE`, and `DELETE` directly from Python. This enables you to manipulate data within a PostgreSQL database through your Python code.

3. **Interfacing with SQLAlchemy**
- SQLAlchemy is an Object-Relational Mapping (ORM) library that provides a higher-level interface for interacting with databases. When using SQLAlchemy with PostgreSQL, `psycopg2` is typically used as the `database driver`.
- SQLAlchemy requires a database driver (such as `psycopg2` for PostgreSQL) to handle the low-level details of communicating with the database, allowing SQLAlchemy to provide a more abstract and `Pythonic` way to work with databases.

4. **Performance and Compatibility**
`psycopg2` is widely used and well-supported, making it a reliable choice for PostgreSQL database interactions. It is designed for performance and efficiency, supporting features like server-side cursors, connection pooling, and asynchronous notifications.

5. **Support for PostgreSQL-Specific Features**
`psycopg2` provides support for many PostgreSQL-specific features, such as advanced data types (like `JSON` and `HSTORE`), custom data types, and asynchronous notifications, which are essential for fully utilizing PostgreSQL’s capabilities.

In summary, `psycopg2` is needed to establish a connection between Python applications and PostgreSQL databases, allowing Python code to execute SQL commands and leverage PostgreSQL’s full feature set. It serves as the necessary driver for SQLAlchemy to interact with PostgreSQL, enabling developers to work with the database in a more Pythonic and abstracted way.

#### How to solve the problem?
Install `psycopg2` using the following command in the designated Python environment:
```bash
pip install psycopg2-binary
```

Verify the installation
```bash
pip show psycopg2-binary
```

Restart VS Code:\
Cmd + Shift + P > Reload Window

### OperationalError
![alt text](images/operationalerror_hostname.png)

#### Understanding the Error:
![alt text](images/hostname_pgadmin4.png)

#### Hot to solve the problem?
Use `localhost` instead of `postgres` as the hostname, with the correct port mapping (5433:5432, so use 5433):
```python
# syntax:
engine = create_engine('postgresql://<username>:<password>@<hostname>:<port>/<dbname>')
```

## Loading data

In [4]:
import pandas as pd

### from JSON data

In [6]:
data = {
    'year': [2020, 2021],
    'category': ['Physics', 'Chemistry'],
    'laureate_type': ['Individual', 'Organization'],
    'birth_date': ['1970-01-01', '1980-01-01'],
    'full_name': ['John Doe', 'Jane Smith'],
    'organization_name': ['Org1', 'Org2'],
    'country': ['USA', 'UK']
}
# in each key-value pair, values are lists with the same length
df_data = pd.DataFrame(data)
df_data

year   category laureate_type  birth_date   full_name organization_name  \
0  2020    Physics    Individual  1970-01-01    John Doe              Org1   
1  2021  Chemistry  Organization  1980-01-01  Jane Smith              Org2   

  country  
0     USA  
1      UK

### from a text file

In [7]:
# by reading from a tsv file
chipot_df = pd.read_csv("files/chipotle.tsv", sep="\t")  # or pd.read_csv("files/chipotle.tsv", delimiter="\t"), `sep` and `delimiter` can be used interchangeably
# `tsv` is  tab separated file. By running `pd.read_csv("files/chipotle.tsv")`, it will return an error. It's required to specify the separator.

print(type(chipot_df))
chipot_df.head()

<class 'pandas.core.frame.DataFrame'>


order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [9]:
# by reading from a csv file
df_data = pd.read_csv('files/nobel_prize_data.csv')
df_data.head()

year    category                                           prize  \
0  1901   Chemistry               The Nobel Prize in Chemistry 1901   
1  1901  Literature              The Nobel Prize in Literature 1901   
2  1901    Medicine  The Nobel Prize in Physiology or Medicine 1901   
3  1901       Peace                      The Nobel Peace Prize 1901   
4  1901       Peace                      The Nobel Peace Prize 1901   

                                          motivation prize_share  \
0  "in recognition of the extraordinary services ...         1/1   
1  "in special recognition of his poetic composit...         1/1   
2  "for his work on serum therapy, especially its...         1/1   
3                                                NaN         1/2   
4                                                NaN         1/2   

  laureate_type                     full_name  birth_date         birth_city  \
0    Individual  Jacobus Henricus van 't Hoff  1852-08-30          Rotterdam   
1    Individual               Sully Prudhomme  1839-03-16              Paris   
2    Individual        Emil Adolf von Behring  1854-03-15  Hansdorf (Lawice)   
3    Individual                Frédéric Passy  1822-05-20              Paris   
4    Individual             Jean Henry Dunant  1828-05-08             Geneva   

      birth_country birth_country_current   sex   organization_name  \
0       Netherlands           Netherlands  Male   Berlin University   
1            France                France  Male                 NaN   
2  Prussia (Poland)                Poland  Male  Marburg University   
3            France                France  Male                 NaN   
4       Switzerland           Switzerland  Male                 NaN   

  organization_city organization_country  ISO  
0            Berlin              Germany  NLD  
1               NaN                  NaN  FRA  
2           Marburg              Germany  POL  
3               NaN                  NaN  FRA  
4               NaN                  NaN  CHE

### from a database using SQLAlchemy

In [14]:
from sqlalchemy import create_engine
import sys
import os

# Add the root directory (two levels up) to the Python path
sys.path.append(os.path.abspath('../../'))

# now, we can import the variable from config.py
from config import POSTGRES_URL_MV

# show the imported variable
print(POSTGRES_URL_MV)

# create  SQLAlchemy engine to connect to the local PostgreSQL database
engine = create_engine(POSTGRES_URL_MV)

# define the SQL query to retrieve data from the database
query = "SELECT * FROM actor a inner join film_actor fa on a.actor_id = fa.actor_id;"

# read the data from the PostgreSQL database into a DataFrame
mv_df = pd.read_sql(query, con=engine)  # `con` stands for connection, it's a database connection string

# print the dataframe
mv_df.head()

postgresql://my_postgres:MyLocalPostgresPassword@localhost:5433/dvdrental


actor_id first_name last_name             last_update  actor_id  film_id  \
0         1   Penelope   Guiness 2013-05-26 14:47:57.620         1        1   
1         1   Penelope   Guiness 2013-05-26 14:47:57.620         1       23   
2         1   Penelope   Guiness 2013-05-26 14:47:57.620         1       25   
3         1   Penelope   Guiness 2013-05-26 14:47:57.620         1      106   
4         1   Penelope   Guiness 2013-05-26 14:47:57.620         1      140   

          last_update  
0 2006-02-15 10:05:03  
1 2006-02-15 10:05:03  
2 2006-02-15 10:05:03  
3 2006-02-15 10:05:03  
4 2006-02-15 10:05:03

NOTE: More about `sys.path` [here](../../python.ipynb#Import-modules-from-different-hierarchy).

### Important Note
Accessing the database using Python can be slow if the database is tremendously large (some data has like millions of rows). It is strongly recommended we first reduce it (by using SQL queries directly in the database) before moving it into Python.

## Working with Dataframes

### Show information about the DataFrame

In [69]:
chipot_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


This method is generally used to check for columns having missing values.

We see that maximum rows are 4622. Columns having less than 4622 value counts definitely have NaN values.

### Show summary statistics

In [70]:
chipot_df.describe()

order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

### Check number of rows and columns

In [20]:
# number of rows
print(df_data.shape[0])

# number of columns
print(df_data.shape[1])

962
16


### Check all headers of the dataframe

In [21]:
df_data.columns

Index(['year', 'category', 'prize', 'motivation', 'prize_share',
       'laureate_type', 'full_name', 'birth_date', 'birth_city',
       'birth_country', 'birth_country_current', 'sex', 'organization_name',
       'organization_city', 'organization_country', 'ISO'],
      dtype='object')

### Check for duplicates and missing values

In [ ]:
# check duplicates
df_data.duplicated()

# check if missing values
df_data.isna()

#### Convert a dataframe to a numpy array

In [26]:
df_data.duplicated().values  # note: .values is an attribute, not a method

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

#### Check if any values in a Series or DataFrame are True

In [27]:
df_data.isna().any()  # check each row in the dataframe
df_data.isna().values.any()  # convert the dataframe to a numpy array then check each values in the array
# both methods work

np.True_

#### Check the number of duplicated and missing values in each column with `.sum()` method

In [28]:
df_data.duplicated().sum()
df_data.isna().sum()

year                       0
category                   0
prize                      0
motivation                88
prize_share                0
laureate_type              0
full_name                  0
birth_date                28
birth_city                31
birth_country             28
birth_country_current     28
sex                       28
organization_name        255
organization_city        255
organization_country     254
ISO                       28
dtype: int64

### Index

In [71]:
chipot_df.index
# it shows that the index is a RangeIndex or Numeric Index

RangeIndex(start=0, stop=4622, step=1)

In [72]:
# retrieve the index values
chipot_df.index.values

array([   0,    1,    2, ..., 4619, 4620, 4621])

By default, indices in pandas are integers, starting from 0. If we want to use a custom index, we need to predefine them first then apply methods as follow:

#### Using `.set_index(<index_column_name>)`:

In [47]:
import pandas as pd
# Sample DataFrame with indices as part of the data
data = {
    'CustomIndex': ['row1', 'row2', 'row3', 'row4'],
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'City': ['New York', 'Los Angeles', 'New York', 'Chicago']
}

# Create DataFrame and set the predefined 'CustomIndex' column as the index
df_customer = pd.DataFrame(data).set_index('CustomIndex')
df_customer

Name         City
CustomIndex                      
row1           Alice     New York
row2             Bob  Los Angeles
row3         Charlie     New York
row4           David      Chicago

NOTE: We can also use the `set_index('CustomIndex')` method after creating the dataframe.

In [49]:
# Create DataFrame 
df_customer = pd.DataFrame(data)

# Set 'CustomIndex' column as the index
df_customer = df_customer.set_index('CustomIndex')
df_customer

Name         City
CustomIndex                      
row1           Alice     New York
row2             Bob  Los Angeles
row3         Charlie     New York
row4           David      Chicago

#### Use .DataFrame's `index` param *for date indices*:

In [51]:
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'City': ['New York', 'Los Angeles', 'New York', 'Chicago']
}

dates = pd.date_range(start='2025-01-01', periods=len(data['Name']), freq='D')

df_customer = pd.DataFrame(data, index=dates)
df_customer

Name         City
2025-01-01    Alice     New York
2025-01-02      Bob  Los Angeles
2025-01-03  Charlie     New York
2025-01-04    David      Chicago

#### Use .DataFrame's `index` param *for string indices*:

In [53]:
data = {
       'Name': ['Alice', 'Kaylee', 'Sophia', 'Lyly'],
       'City': ['New York', 'Los Angeles', 'New York', 'Chicago']
}

custom_index = ['customIndex' + str(i) for i in range(len(data['Name']))]

df_city = pd.DataFrame(data)
df_city.index = custom_index  ## index is a parameter of the .DataFrame() method, we can specify it like this

df_city

Name         City
customIndex0   Alice     New York
customIndex1  Kaylee  Los Angeles
customIndex2  Sophia     New York
customIndex3    Lyly      Chicago

NOTE: 

The `index` param of the `.DataFrame()` method can accept several types of inputs, including:
- List or Array: Simple lists or arrays of values (as above)
- DatetimeIndex: A pandas `DatetimeIndex` object (as above)
- MultiIndex: A pandas `MultiIndex` object

In [54]:
custom_index = pd.MultiIndex.from_product([['Group1', 'Group2'], ['SubGroup1', 'SubGroup2']], names=['Group', 'SubGroup'])
df_city.index = custom_index

df_city

Name         City
Group  SubGroup                      
Group1 SubGroup1   Alice     New York
       SubGroup2  Kaylee  Los Angeles
Group2 SubGroup1  Sophia     New York
       SubGroup2    Lyly      Chicago

#### CategoricalIndex: A pandas' `CategoricalIndex` object

In [55]:
custom_index = pd.CategoricalIndex(['Category1', 'Category2', 'Category1', 'Category2'], categories=['Category1', 'Category2'])
df_city.index = custom_index

df_city

Name         City
Category1   Alice     New York
Category2  Kaylee  Los Angeles
Category1  Sophia     New York
Category2    Lyly      Chicago

## Filtering Dataframes

### Filtering Rows 

#### Using `[]` (boolean indexing)

In [76]:
# find all the rows where the order is a 'Chicken Salad Bowl'
chipot_df[chipot_df['item_name'] == "Chicken Salad Bowl"]

order_id  quantity           item_name  \
44          20         1  Chicken Salad Bowl   
135         60         2  Chicken Salad Bowl   
213         94         2  Chicken Salad Bowl   
256        111         1  Chicken Salad Bowl   
315        137         2  Chicken Salad Bowl   
...        ...       ...                 ...   
4561      1813         2  Chicken Salad Bowl   
4580      1822         1  Chicken Salad Bowl   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  
44    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   
135   [Tomatillo Green Chili Salsa, [Sour Cream, Che...    $22.50   
213   [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $22.50   
256   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75   
315             [Fresh Tomato Salsa, Fajita Vegetables]    $17.50   
...                                                 ...        ...  
4561  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $17.50   
4580  [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...    $11.25   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[110 rows x 5 columns]

#### Using `.filter` (filter based on labels only, not values!)

In [31]:
## assumed that we have a data value:
data = {
    'A': [1, 2, 3],
    'B': [4, 5, 6]
}
## create a dataframe
df_data_2 = pd.DataFrame(data)
df_data_2

A  B
0  1  4
1  2  5
2  3  6

Using `filter()` to filter rows by <span style="color: orange;">index label</span>:

In [32]:
df_data_2.filter(items=[0, 2], axis=0)  # the `items` param is omissible: df_data.filter([0,2], axis=0)

A  B
0  1  4
2  3  6

NOTE: 

`filter()` is generally used to filter <span style="color:orange;">columns</span> based on patterns. It's rarely used to filter rows. For row filtering, the `.filter()` function can only filter based on the index label, not record values, i.e., `df_data['category'].filter(regex='^Phys', axis=0)` seems be be legit, but it will not work.


#### Using `.loc[]` (label-based indexing)

In [46]:
chipot_df.loc[[0, 4, 7]]  # select rows by index

order_id  quantity                     item_name  \
0         1         1  Chips and Fresh Tomato Salsa   
4         2         2                  Chicken Bowl   
7         4         1                 Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75

In [33]:
df_data_2.loc[df_data.birth_date.isna()]

Empty DataFrame
Columns: [A, B]
Index: []

In [34]:
## filter rows where `birth_date` is NaN and return specified columns, i.e., filter rows and columns at the same time
col_subset = ['year', 'category', 'laureate_type', 'birth_date', 'full_name', 'organization_name']

df_data.loc[df_data.birth_date.isna(), col_subset]

year category laureate_type birth_date  \
24   1904    Peace  Organization        NaN   
60   1910    Peace  Organization        NaN   
89   1917    Peace  Organization        NaN   
200  1938    Peace  Organization        NaN   
215  1944    Peace  Organization        NaN   
237  1947    Peace  Organization        NaN   
238  1947    Peace  Organization        NaN   
283  1954    Peace  Organization        NaN   
348  1963    Peace  Organization        NaN   
349  1963    Peace  Organization        NaN   
366  1965    Peace  Organization        NaN   
399  1969    Peace  Organization        NaN   
479  1977    Peace  Organization        NaN   
523  1981    Peace  Organization        NaN   
558  1985    Peace  Organization        NaN   
588  1988    Peace  Organization        NaN   
659  1995    Peace  Organization        NaN   
682  1997    Peace  Organization        NaN   
703  1999    Peace  Organization        NaN   
730  2001    Peace  Organization        NaN   
778  2005    Peace  Organization        NaN   
788  2006    Peace  Organization        NaN   
801  2007    Peace  Organization        NaN   
860  2012    Peace  Organization        NaN   
873  2013    Peace  Organization        NaN   
897  2015    Peace  Organization        NaN   
919  2017    Peace  Organization        NaN   
958  2020    Peace  Organization        NaN   

                                             full_name organization_name  
24   Institut de droit international (Institute of ...               NaN  
60   Bureau international permanent de la Paix (Per...               NaN  
89   Comité international de la Croix Rouge (Intern...               NaN  
200  Office international Nansen pour les Réfugiés ...               NaN  
215  Comité international de la Croix Rouge (Intern...               NaN  
237   American Friends Service Committee (The Quakers)               NaN  
238              Friends Service Council (The Quakers)               NaN  
283  Office of the United Nations High Commissioner...               NaN  
348  Comité international de la Croix Rouge (Intern...               NaN  
349  Ligue des Sociétés de la Croix-Rouge (League o...               NaN  
366            United Nations Children's Fund (UNICEF)               NaN  
399         International Labour Organization (I.L.O.)               NaN  
479                              Amnesty International               NaN  
523  Office of the United Nations High Commissioner...               NaN  
558  International Physicians for the Prevention of...               NaN  
588                 United Nations Peacekeeping Forces               NaN  
659   Pugwash Conferences on Science and World Affairs               NaN  
682     International Campaign to Ban Landmines (ICBL)               NaN  
703                           Médecins Sans Frontières               NaN  
730                              United Nations (U.N.)               NaN  
778          International Atomic Energy Agency (IAEA)               NaN  
788                                       Grameen Bank               NaN  
801   Intergovernmental Panel on Climate Change (IPCC)               NaN  
860                                European Union (EU)               NaN  
873  Organisation for the Prohibition of Chemical W...               NaN  
897                          National Dialogue Quartet               NaN  
919  International Campaign to Abolish Nuclear Weap...               NaN  
958                         World Food Programme (WFP)               NaN

In [37]:
# More examples of `.loc[]` indexer:

# example with labels or indices
df_data_2.loc[0:2, ['A', 'B']]

A  B
0  1  4
1  2  5
2  3  6

In [38]:
# example with boolean conditions
df_data_2.loc[df_data_2['A'] > 1, ['A', 'B']]

A  B
1  2  5
2  3  6

#### Using `.str.contains()`

As mentioned above, `filter()` method only works if we intend to filter COLUMNS based on patterns or filter ROWS based on the index label column. `.str.contains()` enables row filtering:

`.str` is a `Special Accessor` for String operations on Series Data. It allows us to apply `string methods` directly to each element in a Pandas Series containing string values, enabling us to manipulate and analyze text data.

`df_data['category'].str` will return a string accessor type which will cause a KeyError. `.str` cannot be used as an indexer by itself, it has to be followed by string method(s) or condition(s), for example: .str.contains('some_string'); `.str.startswith('some_string')`

In [40]:
filtered_df = df_data[df_data['category'].str.contains('Physics')]
filtered_df


year category                            prize  \
5    1901  Physics  The Nobel Prize in Physics 1901   
11   1902  Physics  The Nobel Prize in Physics 1902   
12   1902  Physics  The Nobel Prize in Physics 1902   
17   1903  Physics  The Nobel Prize in Physics 1903   
18   1903  Physics  The Nobel Prize in Physics 1903   
..    ...      ...                              ...   
948  2019  Physics  The Nobel Prize in Physics 2019   
949  2019  Physics  The Nobel Prize in Physics 2019   
959  2020  Physics  The Nobel Prize in Physics 2020   
960  2020  Physics  The Nobel Prize in Physics 2020   
961  2020  Physics  The Nobel Prize in Physics 2020   

                                            motivation prize_share  \
5    "in recognition of the extraordinary services ...         1/1   
11   "in recognition of the extraordinary service t...         1/2   
12   "in recognition of the extraordinary service t...         1/2   
17   "in recognition of the extraordinary services ...         1/2   
18   "in recognition of the extraordinary services ...         1/4   
..                                                 ...         ...   
948  “for theoretical discoveries in physical cosmo...         1/2   
949  “for the discovery of an exoplanet orbiting a ...         1/4   
959  “for the discovery of a supermassive compact o...         1/4   
960  “for the discovery of a supermassive compact o...         1/4   
961  “for the discovery that black hole formation i...         1/2   

    laureate_type                    full_name  birth_date  \
5      Individual       Wilhelm Conrad Röntgen  1845-03-27   
11     Individual       Hendrik Antoon Lorentz  1853-07-18   
12     Individual                Pieter Zeeman  1865-05-25   
17     Individual      Antoine Henri Becquerel  1852-12-15   
18     Individual  Marie Curie, née Sklodowska  1867-11-07   
..            ...                          ...         ...   
948    Individual                James Peebles  1935-04-25   
949    Individual                 Michel Mayor  1942-01-12   
959    Individual                  Andrea Ghez  1965-06-16   
960    Individual              Reinhard Genzel  1952-03-24   
961    Individual                Roger Penrose  1931-08-08   

                   birth_city             birth_country  \
5          Lennep (Remscheid)         Prussia (Germany)   
11                     Arnhem               Netherlands   
12                 Zonnemaire               Netherlands   
17                      Paris                    France   
18                     Warsaw   Russian Empire (Poland)   
..                        ...                       ...   
948                  Winnipeg                    Canada   
949                  Lausanne               Switzerland   
959              New York, NY  United States of America   
960  Bad Homburg vor der Höhe                   Germany   
961                Colchester            United Kingdom   

        birth_country_current     sex         organization_name  \
5                     Germany    Male         Munich University   
11                Netherlands    Male         Leiden University   
12                Netherlands    Male      Amsterdam University   
17                     France    Male       École Polytechnique   
18                     Poland  Female                       NaN   
..                        ...     ...                       ...   
948                    Canada    Male      Princeton University   
949               Switzerland    Male      University of Geneva   
959  United States of America  Female  University of California   
960                   Germany    Male  University of California   
961            United Kingdom    Male      University of Oxford   

    organization_city      organization_country  ISO  
5              Munich                   Germany  DEU  
11             Leiden               Netherlands  NLD  
12          Amsterdam               Netherlands  NLD  
17              Paris   

In [41]:
filtered_df = df_data[df_data['category'].str.startswith('Phy')]
filtered_df

year category                            prize  \
5    1901  Physics  The Nobel Prize in Physics 1901   
11   1902  Physics  The Nobel Prize in Physics 1902   
12   1902  Physics  The Nobel Prize in Physics 1902   
17   1903  Physics  The Nobel Prize in Physics 1903   
18   1903  Physics  The Nobel Prize in Physics 1903   
..    ...      ...                              ...   
948  2019  Physics  The Nobel Prize in Physics 2019   
949  2019  Physics  The Nobel Prize in Physics 2019   
959  2020  Physics  The Nobel Prize in Physics 2020   
960  2020  Physics  The Nobel Prize in Physics 2020   
961  2020  Physics  The Nobel Prize in Physics 2020   

                                            motivation prize_share  \
5    "in recognition of the extraordinary services ...         1/1   
11   "in recognition of the extraordinary service t...         1/2   
12   "in recognition of the extraordinary service t...         1/2   
17   "in recognition of the extraordinary services ...         1/2   
18   "in recognition of the extraordinary services ...         1/4   
..                                                 ...         ...   
948  “for theoretical discoveries in physical cosmo...         1/2   
949  “for the discovery of an exoplanet orbiting a ...         1/4   
959  “for the discovery of a supermassive compact o...         1/4   
960  “for the discovery of a supermassive compact o...         1/4   
961  “for the discovery that black hole formation i...         1/2   

    laureate_type                    full_name  birth_date  \
5      Individual       Wilhelm Conrad Röntgen  1845-03-27   
11     Individual       Hendrik Antoon Lorentz  1853-07-18   
12     Individual                Pieter Zeeman  1865-05-25   
17     Individual      Antoine Henri Becquerel  1852-12-15   
18     Individual  Marie Curie, née Sklodowska  1867-11-07   
..            ...                          ...         ...   
948    Individual                James Peebles  1935-04-25   
949    Individual                 Michel Mayor  1942-01-12   
959    Individual                  Andrea Ghez  1965-06-16   
960    Individual              Reinhard Genzel  1952-03-24   
961    Individual                Roger Penrose  1931-08-08   

                   birth_city             birth_country  \
5          Lennep (Remscheid)         Prussia (Germany)   
11                     Arnhem               Netherlands   
12                 Zonnemaire               Netherlands   
17                      Paris                    France   
18                     Warsaw   Russian Empire (Poland)   
..                        ...                       ...   
948                  Winnipeg                    Canada   
949                  Lausanne               Switzerland   
959              New York, NY  United States of America   
960  Bad Homburg vor der Höhe                   Germany   
961                Colchester            United Kingdom   

        birth_country_current     sex         organization_name  \
5                     Germany    Male         Munich University   
11                Netherlands    Male         Leiden University   
12                Netherlands    Male      Amsterdam University   
17                     France    Male       École Polytechnique   
18                     Poland  Female                       NaN   
..                        ...     ...                       ...   
948                    Canada    Male      Princeton University   
949               Switzerland    Male      University of Geneva   
959  United States of America  Female  University of California   
960                   Germany    Male  University of California   
961            United Kingdom    Male      University of Oxford   

    organization_city      organization_country  ISO  
5              Munich                   Germany  DEU  
11             Leiden               Netherlands  NLD  
12          Amsterdam               Netherlands  NLD  
17              Paris   

In [42]:
filtered_df = df_data[df_data['category'].str.endswith('try')]
filtered_df

year   category                              prize  \
0    1901  Chemistry  The Nobel Prize in Chemistry 1901   
6    1902  Chemistry  The Nobel Prize in Chemistry 1902   
13   1903  Chemistry  The Nobel Prize in Chemistry 1903   
20   1904  Chemistry  The Nobel Prize in Chemistry 1904   
26   1905  Chemistry  The Nobel Prize in Chemistry 1905   
..    ...        ...                                ...   
936  2019  Chemistry  The Nobel Prize in Chemistry 2019   
937  2019  Chemistry  The Nobel Prize in Chemistry 2019   
938  2019  Chemistry  The Nobel Prize in Chemistry 2019   
950  2020  Chemistry  The Nobel Prize in Chemistry 2020   
951  2020  Chemistry  The Nobel Prize in Chemistry 2020   

                                            motivation prize_share  \
0    "in recognition of the extraordinary services ...         1/1   
6    "in recognition of the extraordinary services ...         1/1   
13   "in recognition of the extraordinary services ...         1/1   
20   "in recognition of his services in the discove...         1/1   
26   "in recognition of his services in the advance...         1/1   
..                                                 ...         ...   
936     “for the development of lithium-ion batteries”         1/3   
937     “for the development of lithium-ion batteries”         1/3   
938     “for the development of lithium-ion batteries”         1/3   
950  “for the development of a method for genome ed...         1/2   
951  “for the development of a method for genome ed...         1/2   

    laureate_type                                  full_name  birth_date  \
0      Individual               Jacobus Henricus van 't Hoff  1852-08-30   
6      Individual                       Hermann Emil Fischer  1852-10-09   
13     Individual                    Svante August Arrhenius  1859-02-19   
20     Individual                         Sir William Ramsay  1852-10-02   
26     Individual  Johann Friedrich Wilhelm Adolf von Baeyer  1835-10-31   
..            ...                                        ...         ...   
936    Individual                              Akira Yoshino  1948-01-30   
937    Individual                            John Goodenough  1922-07-25   
938    Individual                     M. Stanley Whittingham  1941-12-22   
950    Individual                     Emmanuelle Charpentier  1968-12-11   
951    Individual                         Jennifer A. Doudna  1964-02-19   

          birth_city             birth_country     birth_country_current  \
0          Rotterdam               Netherlands               Netherlands   
6         Euskirchen         Prussia (Germany)                   Germany   
13               Vik                    Sweden                    Sweden   
20           Glasgow                  Scotland            United Kingdom   
26            Berlin         Prussia (Germany)                   Germany   
..               ...                       ...                       ...   
936            Suita                     Japan                     Japan   
937             Jena                   Germany                   Germany   
938       Nottingham            United Kingdom            United Kingdom   
950  Juvisy-sur-Orge                    France                    France   
951   Washington, DC  United States of America  United States of America   

        sex                                  organization_name  \
0      Male                                  Berlin University   
6      Male                                  Berlin University   
13     Male                               Stockholm University   
20     Male                          University College London   
26     Male                                  Munich University   
..      ...                                                ...   
936    Male                            Asahi Kasei Corporation   
937    Male                                University of Texas   
938    Male  Binghamton University 

In [43]:
## Extra: Operations on rows using `.str`
df_data['category_upper'] = df_data['category'].str.upper()  # convert each value in the category column to uppercase
df_data['category_replaced'] = df_data['category'].str.replace('Physics', 'Science')  # replace each value in the category column

#### With multiple conditions using `&`

In [77]:
# find all the rows where the quantity is 2 and the order is a 'Chicken Salad Bowl'
chipot_df[(chipot_df['quantity'] == 2) & (chipot_df['item_name'] == "Chicken Salad Bowl")]

order_id  quantity           item_name  \
135         60         2  Chicken Salad Bowl   
213         94         2  Chicken Salad Bowl   
315        137         2  Chicken Salad Bowl   
591        243         2  Chicken Salad Bowl   
633        261         2  Chicken Salad Bowl   
816        337         2  Chicken Salad Bowl   
864        357         2  Chicken Salad Bowl   
1110       459         2  Chicken Salad Bowl   
1923       777         2  Chicken Salad Bowl   
2419       960         2  Chicken Salad Bowl   
3167      1265         2  Chicken Salad Bowl   
3507      1409         2  Chicken Salad Bowl   
4561      1813         2  Chicken Salad Bowl   

                                     choice_description item_price  
135   [Tomatillo Green Chili Salsa, [Sour Cream, Che...    $22.50   
213   [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $22.50   
315             [Fresh Tomato Salsa, Fajita Vegetables]    $17.50   
591   [Tomatillo Green Chili Salsa, [Black Beans, So...    $22.50   
633   [Tomatillo Green Chili Salsa, [Black Beans, Ch...    $22.50   
816   [Tomatillo Green Chili Salsa, [Black Beans, Ch...    $22.50   
864   [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...    $22.50   
1110  [Fresh Tomato Salsa, [Black Beans, Cheese, Gua...    $22.50   
1923  [Tomatillo Green Chili Salsa, [Black Beans, Ch...    $22.50   
2419  [Tomatillo Green Chili Salsa, [Black Beans, Ch...    $22.50   
3167  [Fresh Tomato Salsa, [Pinto Beans, Cheese, Let...    $17.50   
3507  [Tomatillo Green Chili Salsa, [Black Beans, Ch...    $22.50   
4561  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $17.50

### Filtering Columns

#### Using `[]`

In [ ]:
df_data[['column1', 'column2']] 

NOTE: 

Single column (`df_data['column1']`) will work, but` df_data['column1', 'column2']` will not. We have to pass a list of column names to `df_data[]` if there are more than one column.

#### Using `.filter` (filter based on patterns)

In [ ]:
df_data.filter(items=['column1', 'column2']) 

`filter()` function is more flexible if we want to apply pattern filtering criteria, namely `like`, `regex`. Example:

In [ ]:
df_data.filter(like='name', axis=1)  # select columns that contain "name". 
df_data.filter(regex='^na', axis=1)  # select columns that match the regular expression. `^na` means start with "na".

NOTE: 

Param `axis` is optional. Value 1 stands for `column`, while 0 stands for `row`. See the `Filter rows` section for more info regarding row filtering.

#### Using `.loc` (filter based on labels or values)

In [45]:
col_subset = ['year', 'category', 'laureate_type', 'birth_date', 'full_name', 'organization_name']
df_data.loc[:,col_subset]

year    category laureate_type  birth_date                     full_name  \
0    1901   Chemistry    Individual  1852-08-30  Jacobus Henricus van 't Hoff   
1    1901  Literature    Individual  1839-03-16               Sully Prudhomme   
2    1901    Medicine    Individual  1854-03-15        Emil Adolf von Behring   
3    1901       Peace    Individual  1822-05-20                Frédéric Passy   
4    1901       Peace    Individual  1828-05-08             Jean Henry Dunant   
..    ...         ...           ...         ...                           ...   
957  2020    Medicine    Individual  1949-07-02              Michael Houghton   
958  2020       Peace  Organization         NaN    World Food Programme (WFP)   
959  2020     Physics    Individual  1965-06-16                   Andrea Ghez   
960  2020     Physics    Individual  1952-03-24               Reinhard Genzel   
961  2020     Physics    Individual  1931-08-08                 Roger Penrose   

            organization_name  
0           Berlin University  
1                         NaN  
2          Marburg University  
3                         NaN  
4                         NaN  
..                        ...  
957     University of Alberta  
958                       NaN  
959  University of California  
960  University of California  
961      University of Oxford  

[962 rows x 6 columns]

NOTE: The below codes do not work:

```python
chipot_df.loc["item_name", "item_price"]  # does not work
chipot_df.loc[["item_name", "item_price"]]  # does not work
```

## Replacing Characters in a Series

In [15]:
chipot_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


We see that `item_price` is an object and not numerical values. To apply mathematical operations on it, we need to convert it to a numerical value as follow:

In [16]:
# in Python, to replace a character with other ones, we can use `replace` function:
"$8.50".replace("$", "")

'8.50'

In [17]:
# However, in pandas, we are dealing with a Series of data. To replace the $ sign with an empty string, we use `.str.replace` instead
chipot_df['item_price'].str.replace('$', '')

0        2.39 
1        3.39 
2        3.39 
3        2.39 
4       16.98 
         ...  
4617    11.75 
4618    11.75 
4619    11.25 
4620     8.75 
4621     8.75 
Name: item_price, Length: 4622, dtype: object

## Converting Data Types

In [81]:
# Convert to float using `astype`
df_data.prize_share.astype(str)  # convert values in a series to string
df_data.scores.astype(float)  # convert to float
df_data.level.astype(int)  # convert to int
df_data.is_alive.astype(bool)  # convert to bool

chipot_df['item_price'].str.replace('$', '').astype(float)

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

`Datetime` conversion is quite different and cannot use the `.astype()` as above.

In [57]:
df_data.birth_date = pd.to_datetime(df_data.birth_date)

type(df_data.birth_date[0])

pandas._libs.tslibs.timestamps.Timestamp

## Column Manipulation

### Split Columns and create a new dataframe

Assumed that the `prize_share` column looks like this:

```plaintext
       category prize_share
0     Chemistry         1/1
1    Literature         1/1
2      Medicine         1/1
3         Peace         1/2
4         Peace         1/2
..          ...         ...
957    Medicine         1/3
958       Peace         1/1
959     Physics         1/4
960     Physics         1/4
961     Physics         1/2

[962 rows x 2 columns]
```

We want to create a new column called `share_pct` which is of type float. To do that, we first need to split the `prize_share` column (which is a string and looks like a fraction '1/2') into two columns using the `str.split()` method:

In [58]:
separated_values = df_data.prize_share.str.split('/', expand=True)
# .split() is a string method, so it must be followed by the `.str` accessor

separated_values

0  1
0    1  1
1    1  1
2    1  1
3    1  2
4    1  2
..  .. ..
957  1  3
958  1  1
959  1  4
960  1  4
961  1  2

[962 rows x 2 columns]

Next, we assign two variables, namely `numerator` and `denominator` for `separated_values[0]` and `separated_values[1]`, respectively. We use the `astype(int)` method to convert the values to integer.

In [59]:
numerator = separated_values[0].astype(int)
denominator = separated_values[1].astype(int)
share_pct = numerator / denominator  # this is a pandas series with the datatype of float

share_pct

0      1.000000
1      1.000000
2      1.000000
3      0.500000
4      0.500000
         ...   
957    0.333333
958    1.000000
959    0.250000
960    0.250000
961    0.500000
Length: 962, dtype: float64

#### `expand` parameter of the `str.split()` method:

By default, the `expand` value is False. Compare the outputs below:

- output from `expand=False`:
```plaintext
0      [1, 1]
1      [1, 1]
2      [1, 1]
3      [1, 2]
4      [1, 2]
        ...  
957    [1, 3]
958    [1, 1]
959    [1, 4]
960    [1, 4]
961    [1, 2]
Name: prize_share, Length: 962, dtype: object
<class 'pandas.core.series.Series'>
```

We see that there is only one column. Values in each row are lists.


- output from `expand=True`:
```plaintext
     0  1
0    1  1
1    1  1
2    1  1
3    1  2
4    1  2
..  .. ..
957  1  3
958  1  1
959  1  4
960  1  4
961  1  2

[962 rows x 2 columns]
<class 'pandas.core.frame.DataFrame'>
```
We see that there are two columns.

### Convert a string column to Fraction dtype

Simply creating a new `share_pct` column, and we need to get hassled creating other temporary columns (`numerator`, `denominator`) and need to have many more steps before achieving the result we want. Isn't there a faster way to do this? Actually, there is.

Another method is that, we can first convert the `prize_share` column, which values are strings but look like fractions, into fractions dtype, then convert the fractions to float.

Convert to fraction dtype:

In [73]:
from fractions import Fraction

df_data.prize_share = df_data.prize_share.apply(lambda x: float(Fraction(x)))

df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 962 entries, 0 to 961
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   year                   962 non-null    int64         
 1   category               962 non-null    object        
 2   prize                  962 non-null    object        
 3   motivation             874 non-null    object        
 4   prize_share            962 non-null    float64       
 5   share_pct              962 non-null    float64       
 6   laureate_type          962 non-null    object        
 7   full_name              962 non-null    object        
 8   birth_date             934 non-null    datetime64[ns]
 9   birth_city             931 non-null    object        
 10  birth_country          934 non-null    object        
 11  birth_country_current  934 non-null    object        
 12  sex                    934 non-null    object        
 13  organ

Now the `prize_share` column is of type float, and we no longer need to create a new `share_pct` column. We could analyze data directly using the `prize_share` column.

## Add a new column to pandas dataset 

### Using direct assignment

In [61]:
import pandas as pd

data = {
       'state': ['CA', 'NY', 'TX', 'CA', 'TX', 'NY', 'TX', 'CA'],
       'count': [1, 2, 3, 2, 1, 1, 2, 1]
}

df_state = pd.DataFrame(data)

# a series to be added to the dataframe
new_int_col = [111, 222, 333, 444, 555, 666, 777, 888]

# add a new column called `new_int_col` to the dataframe
df_state['new_int_col'] = new_int_col

df_state

state  count  new_int_col
0    CA      1          111
1    NY      2          222
2    TX      3          333
3    CA      2          444
4    TX      1          555
5    NY      1          666
6    TX      2          777
7    CA      1          888

Note that if the length of `new_int_col` is not equal to that of `df.index`, a ValueError will be raised:

```plaintext
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-70-bec12d117801> in <cell line: 20>()
     18 # Add a new column with full state name
     19 df['full_state'] = df['state'].map(state_names)
---> 20 df['test_col'] = [111, 222, 333, 444, 555, 666, 777]
     21 print(df)
     22 

3 frames
/usr/local/lib/python3.10/dist-packages/pandas/core/common.py in require_length_match(data, index)
    574     """
    575     if len(data) != len(index):
--> 576         raise ValueError(
    577             "Length of values "
    578             f"({len(data)}) "

ValueError: Length of values (7) does not match length of index (8)
```

To avoid that, first convert the `new_int_col` to a pandas series using `pd.Series()` (see below).

### Using Pandas' `pd.Series()`

In [62]:
# a series to be added to the dataframe
new_int_col = [111, 222, 333, 444, 555, 666, 777] # length: 7 < len(data.state) = 8
new_str_col = ['AAA', 'BBB', 'CCC', 'DDD', 'EEE', 'FFF', 'GGG', 'HHH', 'III']   # length: 9 > len(data.state) = 8

# add a new column called `new_int_col` to the dataframe
df_state['new_int_col'] = pd.Series(new_int_col)
df_state['new_str_col'] = pd.Series(new_str_col)

df_state

state  count  new_int_col new_str_col
0    CA      1        111.0         AAA
1    NY      2        222.0         BBB
2    TX      3        333.0         CCC
3    CA      2        444.0         DDD
4    TX      1        555.0         EEE
5    NY      1        666.0         FFF
6    TX      2        777.0         GGG
7    CA      1          NaN         HHH

Pandas automatically aligns the length of new_int_col and new_str_col to match df.index. If the series is shorter, pandas adds NaN for missing values; if longer, pandas ignores the extra values.

To handle missing values differently, we can use `.reindex(index=df.index, fill_value='Missing')` to fill with a specific value:

In [68]:
new_str_col = ['AAA', 'BBB', 'CCC', 'DDD', 'EEE', 'FFF', 'GGG']

# add a new column called `new_str_col` to the dataframe while customizing the default `NaN` value (use 'Missing' instead)
df_state['new_str_col'] = pd.Series(new_str_col).reindex(index=df_state.index, fill_value='Missing')

df_state

state  count new_str_col
0    CA      1         AAA
1    NY      2         BBB
2    TX      3         CCC
3    CA      2         DDD
4    TX      1         EEE
5    NY      1         FFF
6    TX      2         GGG
7    CA      1     Missing

### Using `.insert()`

In [69]:
# get the column index of the `prize_share` column
prize_share_index = df_data.columns.get_loc('prize_share')  # see more details of .get_loc() in other section below
# .get_loc() is a method of the .columns attribute. Hence to call it, we must call the .columns first.

prize_share_index

4

In [70]:
# add the new share_pct column next to the `prize_share` column
df_data.insert(prize_share_index+1, 'share_pct', share_pct)  # the new `share_pct` column will be located at index 5, a column after the `prize_share` column

df_data

year    category                                           prize  \
0    1901   Chemistry               The Nobel Prize in Chemistry 1901   
1    1901  Literature              The Nobel Prize in Literature 1901   
2    1901    Medicine  The Nobel Prize in Physiology or Medicine 1901   
3    1901       Peace                      The Nobel Peace Prize 1901   
4    1901       Peace                      The Nobel Peace Prize 1901   
..    ...         ...                                             ...   
957  2020    Medicine  The Nobel Prize in Physiology or Medicine 2020   
958  2020       Peace                      The Nobel Peace Prize 2020   
959  2020     Physics                 The Nobel Prize in Physics 2020   
960  2020     Physics                 The Nobel Prize in Physics 2020   
961  2020     Physics                 The Nobel Prize in Physics 2020   

                                            motivation prize_share  share_pct  \
0    "in recognition of the extraordinary services ...         1/1   1.000000   
1    "in special recognition of his poetic composit...         1/1   1.000000   
2    "for his work on serum therapy, especially its...         1/1   1.000000   
3                                                  NaN         1/2   0.500000   
4                                                  NaN         1/2   0.500000   
..                                                 ...         ...        ...   
957           “for the discovery of Hepatitis C virus”         1/3   0.333333   
958  “for its efforts to combat hunger, for its con...         1/1   1.000000   
959  “for the discovery of a supermassive compact o...         1/4   0.250000   
960  “for the discovery of a supermassive compact o...         1/4   0.250000   
961  “for the discovery that black hole formation i...         1/2   0.500000   

    laureate_type                     full_name birth_date  \
0      Individual  Jacobus Henricus van 't Hoff 1852-08-30   
1      Individual               Sully Prudhomme 1839-03-16   
2      Individual        Emil Adolf von Behring 1854-03-15   
3      Individual                Frédéric Passy 1822-05-20   
4      Individual             Jean Henry Dunant 1828-05-08   
..            ...                           ...        ...   
957    Individual              Michael Houghton 1949-07-02   
958  Organization    World Food Programme (WFP)        NaT   
959    Individual                   Andrea Ghez 1965-06-16   
960    Individual               Reinhard Genzel 1952-03-24   
961    Individual                 Roger Penrose 1931-08-08   

                   birth_city             birth_country  \
0                   Rotterdam               Netherlands   
1                       Paris                    France   
2           Hansdorf (Lawice)          Prussia (Poland)   
3                       Paris                    France   
4                      Geneva               Switzerland   
..                        ...                       ...   
957                       NaN            United Kingdom   
958                       NaN                       NaN   
959              New York, NY  United States of America   
960  Bad Homburg vor der Höhe                   Germany   
961                Colchester            United Kingdom   

        birth_country_current     sex         organization_name  \
0                 Netherlands    Male         Berlin University   
1                      France    Male                       NaN   
2                      Poland    Male        Marburg University   
3                      France    Male                       NaN   
4                 Switzerland    Male                       NaN   
..                        ...     ...                       ...   
957            United Kingdom    Male     University of Alberta   
958                       NaN     NaN                       NaN   
959  United States of America  Female  University of California   
960                   Germany  

### Other methods

#### Map an existing column to create a new one

In [71]:
data = {
    'state': ['CA', 'NY', 'TX', 'CA', 'TX', 'NY', 'TX', 'CA'],
    'count': [1, 2, 3, 2, 1, 1, 2, 1]
}

df_state = pd.DataFrame(data)

# create a dictionary called `state_names` that maps the abbreviations to full state names
state_names = {
  'CA': 'California',
  'NY': 'New York',
  'TX': 'Texas'
}

# create a new column called 'full_state' by mapping the 'state' column to the 'state_names' dictionary
df_state['full_state'] = df_state.state.map(state_names)

df_state

state  count  full_state
0    CA      1  California
1    NY      2    New York
2    TX      3       Texas
3    CA      2  California
4    TX      1       Texas
5    NY      1    New York
6    TX      2       Texas
7    CA      1  California

## Sorting

### Sort by index

In [25]:
df_data.sort_index()

year    category                                           prize  \
0    1901   Chemistry               The Nobel Prize in Chemistry 1901   
1    1901  Literature              The Nobel Prize in Literature 1901   
2    1901    Medicine  The Nobel Prize in Physiology or Medicine 1901   
3    1901       Peace                      The Nobel Peace Prize 1901   
4    1901       Peace                      The Nobel Peace Prize 1901   
..    ...         ...                                             ...   
957  2020    Medicine  The Nobel Prize in Physiology or Medicine 2020   
958  2020       Peace                      The Nobel Peace Prize 2020   
959  2020     Physics                 The Nobel Prize in Physics 2020   
960  2020     Physics                 The Nobel Prize in Physics 2020   
961  2020     Physics                 The Nobel Prize in Physics 2020   

                                            motivation prize_share  \
0    "in recognition of the extraordinary services ...         1/1   
1    "in special recognition of his poetic composit...         1/1   
2    "for his work on serum therapy, especially its...         1/1   
3                                                  NaN         1/2   
4                                                  NaN         1/2   
..                                                 ...         ...   
957           “for the discovery of Hepatitis C virus”         1/3   
958  “for its efforts to combat hunger, for its con...         1/1   
959  “for the discovery of a supermassive compact o...         1/4   
960  “for the discovery of a supermassive compact o...         1/4   
961  “for the discovery that black hole formation i...         1/2   

    laureate_type                     full_name  birth_date  \
0      Individual  Jacobus Henricus van 't Hoff  1852-08-30   
1      Individual               Sully Prudhomme  1839-03-16   
2      Individual        Emil Adolf von Behring  1854-03-15   
3      Individual                Frédéric Passy  1822-05-20   
4      Individual             Jean Henry Dunant  1828-05-08   
..            ...                           ...         ...   
957    Individual              Michael Houghton  1949-07-02   
958  Organization    World Food Programme (WFP)         NaN   
959    Individual                   Andrea Ghez  1965-06-16   
960    Individual               Reinhard Genzel  1952-03-24   
961    Individual                 Roger Penrose  1931-08-08   

                   birth_city             birth_country  \
0                   Rotterdam               Netherlands   
1                       Paris                    France   
2           Hansdorf (Lawice)          Prussia (Poland)   
3                       Paris                    France   
4                      Geneva               Switzerland   
..                        ...                       ...   
957                       NaN            United Kingdom   
958                       NaN                       NaN   
959              New York, NY  United States of America   
960  Bad Homburg vor der Höhe                   Germany   
961                Colchester            United Kingdom   

        birth_country_current     sex         organization_name  \
0                 Netherlands    Male         Berlin University   
1                      France    Male                       NaN   
2                      Poland    Male        Marburg University   
3                      France    Male                       NaN   
4                 Switzerland    Male                       NaN   
..                        ...     ...                       ...   
957            United Kingdom    Male     University of Alberta   
958                       NaN     NaN                       NaN   
959  United States of America  Female  University of California   
960                   Germany    Male  University of California   
961            United Kingdom    Male      University of Oxford   

    organization

### Sorting Pandas Series

In [24]:
df_data['year'].sort_values(ascending=False)  # no 'by' parameter for series sorting

961    2020
955    2020
950    2020
951    2020
952    2020
       ... 
2      1901
1      1901
4      1901
5      1901
0      1901
Name: year, Length: 962, dtype: int64

### Sorting Dataframes

In [82]:
# sort by item price in descending order and order_id in ascending order
chipot_df.sort_values(by=['item_price', 'order_id'], ascending=[False, True])

order_id  quantity            item_name  \
607        250         1     Steak Salad Bowl   
1418       576         1  Barbacoa Salad Bowl   
1421       576         1  Barbacoa Salad Bowl   
1825       738         1  Barbacoa Salad Bowl   
1865       756         1  Carnitas Salad Bowl   
...        ...       ...                  ...   
3936      1578         1          Canned Soda   
4001      1602         1        Bottled Water   
4008      1604         1          Canned Soda   
4051      1621         1          Canned Soda   
4069      1629         1        Bottled Water   

                                     choice_description item_price  
607                       [Fresh Tomato Salsa, Lettuce]     $9.39   
1418                         [Roasted Chili Corn Salsa]     $9.39   
1421                         [Roasted Chili Corn Salsa]     $9.39   
1825  [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...     $9.39   
1865   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $9.39   
...                                                 ...        ...  
3936                                  [Diet Dr. Pepper]     $1.09   
4001                                                NaN     $1.09   
4008                                        [Diet Coke]     $1.09   
4051                                           [Sprite]     $1.09   
4069                                                NaN     $1.09   

[4622 rows x 5 columns]

## Group BY

In [83]:
# create a new dataframe to work with
chipot_new_df = chipot_df.loc[:,['item_name', 'quantity', 'item_price']]
chipot_new_df.item_price = chipot_new_df.item_price.str.replace('$', '').astype(float)

chipot_new_df.groupby('item_name', as_index=False).mean()

item_name  quantity  item_price
0                       6 Pack Soft Drink  1.018519    6.610185
1                           Barbacoa Bowl  1.000000   10.187273
2                        Barbacoa Burrito  1.000000    9.832418
3                   Barbacoa Crispy Tacos  1.090909   10.928182
4                     Barbacoa Salad Bowl  1.000000   10.640000
5                     Barbacoa Soft Tacos  1.000000   10.018400
6                           Bottled Water  1.302469    1.867654
7                                    Bowl  2.000000   14.800000
8                                 Burrito  1.000000    7.400000
9                             Canned Soda  1.211538    1.320577
10                      Canned Soft Drink  1.166113    1.457641
11                          Carnitas Bowl  1.044118   10.833971
12                       Carnitas Burrito  1.016949   10.132712
13                  Carnitas Crispy Tacos  1.142857   11.137143
14                         Carnitas Salad  1.000000    8.990000
15                    Carnitas Salad Bowl  1.000000   11.056667
16                    Carnitas Soft Tacos  1.000000    9.398500
17                           Chicken Bowl  1.048209   10.113953
18                        Chicken Burrito  1.068716   10.082857
19                   Chicken Crispy Tacos  1.063830   10.045319
20                          Chicken Salad  1.000000    9.010000
21                     Chicken Salad Bowl  1.118182   11.170455
22                     Chicken Soft Tacos  1.043478    9.635565
23                                  Chips  1.090047    2.342844
24           Chips and Fresh Tomato Salsa  1.181818    3.285091
25                    Chips and Guacamole  1.056367    4.595073
26      Chips and Mild Fresh Tomato Salsa  1.000000    3.000000
27     Chips and Roasted Chili Corn Salsa  1.045455    3.084091
28     Chips and Roasted Chili-Corn Salsa  1.000000    2.390000
29  Chips and Tomatillo Green Chili Salsa  1.046512    3.087209
30    Chips and Tomatillo Red Chili Salsa  1.041667    3.072917
31  Chips and Tomatillo-Green Chili Salsa  1.064516    2.544194
32    Chips and Tomatillo-Red Chili Salsa  1.250000    2.987500
33                           Crispy Tacos  1.000000    7.400000
34                                   Izze  1.000000    3.390000
35                       Nantucket Nectar  1.074074    3.641111
36                                  Salad  1.000000    7.400000
37                          Side of Chips  1.089109    1.840594
38                             Steak Bowl  1.047393   10.711801
39                          Steak Burrito  1.048913   10.465842
40                     Steak Crispy Tacos  1.028571   10.209714
41                            Steak Salad  1.000000    8.915000
42                       Steak Salad Bowl  1.068966   11.847931
43                       Steak Soft Tacos  1.018182    9.746364
44                            Veggie Bowl  1.023529   10.211647
45                         Veggie Burrito  1.021053    9.839684
46                    Veggie Crispy Tacos  1.000000    8.490000
47                           Veggie Salad  1.000000    8.490000
48                      Veggie Salad Bowl  1.000000   10.138889
49                      Veggie Soft Tacos  1.142857   10.565714

### Group Keys
The group_keys=False parameter is used in the `groupby` operation to prevent the group labels from being added back to the result’s index when applying a function.

In [84]:
# without group_keys=False
import pandas as pd

df = pd.DataFrame({
    'category': ['A', 'A', 'B', 'B'],
    'value': [1, 2, 3, 4]
})

result = df.groupby('category').apply(lambda x: x['value'], include_groups=False)
result

category   
A         0    1
          1    2
B         2    3
          3    4
Name: value, dtype: int64

In [85]:
# With group_keys=False
result = df.groupby('category', group_keys=False).apply(lambda x: x['value'], include_groups=False)
result

0    1
1    2
2    3
3    4
Name: value, dtype: int64

## Setting display options

### Float Format

In [ ]:
# thousand separator and 2 decimal places
pd.options.display.float_format = '{:,.2f}'.format

Note: `.format` is required. It specifies that the formatting string should be used as a function to format floating-point numbers.

`{:,.2f}` is a template for formatting numbers. To format a specific number, use it this way: `'{:,.2f}'.format(1234.56789101112)`. 
$$
\implies \text{'1,234.57'}
$$


`{:,.2f}.format` turns the template into a function. Assigning it to `pd.options.display.float_format` makes pandas use it to display all floating-point numbers in the project. 

### Display all rows

In [86]:
pd.set_option("display.max_rows", None) 
chipot_new_df.groupby(['item_name', 'quantity']).mean()

item_price
item_name                             quantity            
6 Pack Soft Drink                     1           6.490000
                                      2          12.980000
Barbacoa Bowl                         1          10.187273
Barbacoa Burrito                      1           9.832418
Barbacoa Crispy Tacos                 1          10.171000
                                      2          18.500000
Barbacoa Salad Bowl                   1          10.640000
Barbacoa Soft Tacos                   1          10.018400
Bottled Water                         1           1.435108
                                      2           2.795000
                                      3           4.090000
                                      4           6.000000
                                      5           7.500000
                                      7          10.500000
                                      10         15.000000
Bowl                                  1           7.400000
                                      3          22.200000
Burrito                               1           7.400000
Canned Soda                           1           1.090000
                                      2           2.180000
                                      4           4.360000
Canned Soft Drink                     1           1.250000
                                      2           2.500000
                                      3           3.750000
                                      4           5.000000
Carnitas Bowl                         1          10.272121
                                      2          23.500000
                                      3          35.250000
Carnitas Burrito                      1           9.988448
                                      2          18.500000
Carnitas Crispy Tacos                 1           9.996667
                                      2          17.980000
Carnitas Salad                        1           8.990000
Carnitas Salad Bowl                   1          11.056667
Carnitas Soft Tacos                   1           9.398500
Chicken Bowl                          1           9.668139
                                      2          18.823226
                                      3          29.595000
Chicken Burrito                       1           9.514549
                                      2          17.722857
                                      3          26.250000
                                      4          35.000000
Chicken Crispy Tacos                  1           9.548864
                                      2          17.326667
Chicken Salad                         1           9.010000
Chicken Salad Bowl                    1           9.806701
                                      2          21.346154
Chicken Soft Tacos                    1           9.247000
                                      2          18.184000
Chips                                 1           2.149171
                                      2           4.300000
                                      3           6.450000
Chips and Fresh Tomato Salsa          1           2.739143
                                      2           5.900000
                                      4          11.800000
                                      15         44.250000
Chips and Guacamole                   1           4.341560
                                      2           8.840000
                                      3          13.350000
                                      4          17.800000
Chips and Mild Fresh Tomato Salsa     1           3.000000
Chips and Roasted Chili Corn Salsa    1           2.950000
                                      2           5.900000
Chips and Roasted Chili-Corn Salsa    1           2.390000
Chips and Tomatillo Green Chili Salsa 1           2.950000
                                      2           5.900000
Chips and Tomatillo Red Chili Salsa 

### Reset the display options

In [87]:
pd.reset_option("display.max_rows")

## Aggregating Dataframes

### Count

#### Use `.value_counts()`

In [74]:
biology = df_data.sex.value_counts()
print(biology)
print(type(biology))

sex
Male      876
Female     58
Name: count, dtype: int64
<class 'pandas.core.series.Series'>


#### Use `groupby()` and `count()`

In [75]:
biology = df_data.groupby('sex')['sex'].count()
print(biology)
print(type(biology))

sex
Female     58
Male      876
Name: sex, dtype: int64
<class 'pandas.core.series.Series'>


We see that the output is exactly the same as `value_counts()`. But using `df_data.groupby('sex')['sex'].count()` is a bit tedious if we only want to show the count of each sex in the `sex` column itself. However, this method is useful when we want to count other columns grouping by a particular column.

For example:

In [76]:
df_data.groupby('sex')[['birth_date', 'birth_city', 'birth_country']].count()

birth_date  birth_city  birth_country
sex                                          
Female          58          58             58
Male           876         873            876

In [78]:
# or use `groupby` and `agg`
df_data.groupby('sex')[['birth_date', 'birth_city', 'birth_country']].agg(['count'])

birth_date birth_city birth_country
            count      count         count
sex                                       
Female         58         58            58
Male          876        873           876

Here, we group by the `sex` column and count the `birth_date`, `birth_city`, and `birth_country` columns.

### Sum

In [88]:
# aggregate on the item_name column and only display the `sum` of the item_price column
chipot_new_df.groupby(['item_name']).sum()['item_price']

# in this case, the output is a series

item_name
6 Pack Soft Drink                         356.95
Barbacoa Bowl                             672.36
Barbacoa Burrito                          894.75
Barbacoa Crispy Tacos                     120.21
Barbacoa Salad Bowl                       106.40
Barbacoa Soft Tacos                       250.46
Bottled Water                             302.56
Bowl                                       29.60
Burrito                                    44.40
Canned Soda                               137.34
Canned Soft Drink                         438.75
Carnitas Bowl                             736.71
Carnitas Burrito                          597.83
Carnitas Crispy Tacos                      77.96
Carnitas Salad                              8.99
Carnitas Salad Bowl                        66.34
Carnitas Soft Tacos                       375.94
Chicken Bowl                             7342.73
Chicken Burrito                          5575.82
Chicken Crispy Tacos                      472.13
Chicken Sa

NOTE:

The above code works because the `chipot_new_df`'s columns are all numerical columns except the `item_name` used for grouping. It's recommended to query the numerical column before aggregating to avoid error in case some other columns are not numerical.

Instead of this: `chipot_new_df.groupby(['item_name']).sum()['item_price']`, do this: `chipot_new_df.groupby(['item_name'])['item_price'].sum()`

### Average

Average is divided into:
- Weighted Average
- Simple Average

#### Weighted Average

**Weighted average** is calculated by summing and dividing, and there’s no built-in pandas function for it—you have to calculate it manually using sums and divisions. This is used when different data points have different levels of importance or frequency. For example, when calculating an average price per item considering the quantity sold, the weighted average is more meaningful. However, it is less frequently used compared to the simple average because it requires additional data (weights) and is more specific to contexts where weighting is necessary.

$$
\text{Weighted Average} = \frac{\sum(\text{Value} \times \text{Weight})}{\sum(\text{Weight})}
$$

#### Simple Average

This is more commonly used in data analysis. It’s straightforward and easy to interpret, making it useful for summarizing data and understanding the central tendency of values. The simple average is appropriate when all data points are equally important.

Example: Let's have a look at an item named: '6 Pack Soft Drink':

In [96]:
chipot_new_df[chipot_new_df.item_name=='6 Pack Soft Drink']

item_name  quantity  item_price
298   6 Pack Soft Drink         1        6.49
341   6 Pack Soft Drink         1        6.49
357   6 Pack Soft Drink         1        6.49
388   6 Pack Soft Drink         1        6.49
417   6 Pack Soft Drink         1        6.49
553   6 Pack Soft Drink         1        6.49
639   6 Pack Soft Drink         1        6.49
721   6 Pack Soft Drink         1        6.49
743   6 Pack Soft Drink         1        6.49
879   6 Pack Soft Drink         1        6.49
1026  6 Pack Soft Drink         1        6.49
1051  6 Pack Soft Drink         1        6.49
1124  6 Pack Soft Drink         1        6.49
1167  6 Pack Soft Drink         1        6.49
1260  6 Pack Soft Drink         1        6.49
1279  6 Pack Soft Drink         1        6.49
1392  6 Pack Soft Drink         1        6.49
1755  6 Pack Soft Drink         1        6.49
1758  6 Pack Soft Drink         1        6.49
1849  6 Pack Soft Drink         1        6.49
1860  6 Pack Soft Drink         1        6.49
1916  6 Pack Soft Drink         1        6.49
1922  6 Pack Soft Drink         1        6.49
1937  6 Pack Soft Drink         1        6.49
1976  6 Pack Soft Drink         1        6.49
2108  6 Pack Soft Drink         1        6.49
2232  6 Pack Soft Drink         1        6.49
2374  6 Pack Soft Drink         1        6.49
2389  6 Pack Soft Drink         1        6.49
2713  6 Pack Soft Drink         1        6.49
2793  6 Pack Soft Drink         1        6.49
2860  6 Pack Soft Drink         1        6.49
3010  6 Pack Soft Drink         1        6.49
3132  6 Pack Soft Drink         1        6.49
3141  6 Pack Soft Drink         1        6.49
3389  6 Pack Soft Drink         2       12.98
3422  6 Pack Soft Drink         1        6.49
3600  6 Pack Soft Drink         1        6.49
3614  6 Pack Soft Drink         1        6.49
3633  6 Pack Soft Drink         1        6.49
3693  6 Pack Soft Drink         1        6.49
3766  6 Pack Soft Drink         1        6.49
3789  6 Pack Soft Drink         1        6.49
3806  6 Pack Soft Drink         1        6.49
3836  6 Pack Soft Drink         1        6.49
3875  6 Pack Soft Drink         1        6.49
3886  6 Pack Soft Drink         1        6.49
4169  6 Pack Soft Drink         1        6.49
4174  6 Pack Soft Drink         1        6.49
4359  6 Pack Soft Drink         1        6.49
4363  6 Pack Soft Drink         1        6.49
4522  6 Pack Soft Drink         1        6.49
4527  6 Pack Soft Drink         1        6.49
4535  6 Pack Soft Drink         1        6.49

We see that:


$$
{\sum(\text{Item Price})} = 6.49 \times 53 + 12.98 = 356.95
$$

$$
{\sum(\text{Weight})} = 1 \times 53 + 2 \times 1 = 55
$$

hence, <span style="color: orange;">Weighted Average:</span>
$$
w\_avg = 356.95 \div 55 = 6.49
$$

Number of rows: 
```python
len(chipot_new_df[chipot_new_df.item_name=='6 Pack Soft Drink'])  # returns 54 rows
```

hence, <span style="color: orange;">Simple Average:</span>
$$
s\_avg = 356.95 \div 54 = 6.61
$$

In a more Pythonic way, the Weighted Average can be calculated as:

In [101]:
# Calculate the Weighted Average per `item_name` by dividing the sum of item_price (total value) by the sum of quantity (total weight)
total_revenue_per_item = chipot_new_df.groupby(by=['item_name']).apply(
    lambda x: (x['item_price']).sum() / x['quantity'].sum(),
    include_groups=False
)
# the above function is exactly the same as: 
# chipot_new_df.groupby(['item_name'])['item_price'].sum() / chipot_new_df.groupby(['item_name'])['quantity'].sum()

total_revenue_per_item

item_name
6 Pack Soft Drink                         6.490000
Barbacoa Bowl                            10.187273
Barbacoa Burrito                          9.832418
Barbacoa Crispy Tacos                    10.017500
Barbacoa Salad Bowl                      10.640000
Barbacoa Soft Tacos                      10.018400
Bottled Water                             1.433934
Bowl                                      7.400000
Burrito                                   7.400000
Canned Soda                               1.090000
Canned Soft Drink                         1.250000
Carnitas Bowl                            10.376197
Carnitas Burrito                          9.963833
Carnitas Crispy Tacos                     9.745000
Carnitas Salad                            8.990000
Carnitas Salad Bowl                      11.056667
Carnitas Soft Tacos                       9.398500
Chicken Bowl                              9.648791
Chicken Burrito                           9.434552
Chicken Crispy Tacos 

In [26]:
# Calculate the Simple Average per `item_name` by finding the .mean() of item_price
chipot_new_df.groupby(['item_name'])['item_price'].mean()

item_name
6 Pack Soft Drink                         6.610185
Barbacoa Bowl                            10.187273
Barbacoa Burrito                          9.832418
Barbacoa Crispy Tacos                    10.928182
Barbacoa Salad Bowl                      10.640000
Barbacoa Soft Tacos                      10.018400
Bottled Water                             1.867654
Bowl                                     14.800000
Burrito                                   7.400000
Canned Soda                               1.320577
Canned Soft Drink                         1.457641
Carnitas Bowl                            10.833971
Carnitas Burrito                         10.132712
Carnitas Crispy Tacos                    11.137143
Carnitas Salad                            8.990000
Carnitas Salad Bowl                      11.056667
Carnitas Soft Tacos                       9.398500
Chicken Bowl                             10.113953
Chicken Burrito                          10.082857
Chicken Crispy Tacos 